aggregator/entry.go

```diff
@@ -0,0 +1,242 @@
+	sync.RWMutex
+
+	opts           Options                         // options
+	lists          *MetricLists                    // metric lists
+	version        int                             // current policy version
+	numWriters     int32                           // number of writers writing to this entry
+	lastAccessInNs int64                           // last access time
+	aggregations   map[policy.Policy]*list.Element // aggregations at different resolutions
+	closed         int32                           // whether the entry is closed
+}
+
+// NewEntry creates a new entry
+func NewEntry(lists *MetricLists, opts Options) *Entry {
+	e := &Entry{
+		opts:         opts,
+		aggregations: make(map[policy.Policy]*list.Element),
+	}
+	e.ResetSetData(lists)
+	return e
+}
+
+// IncWriter increases the writer count
+func (e *Entry) IncWriter() { atomic.AddInt32(&e.numWriters, 1) }
+
+// DecWriter decreases the writer count
+func (e *Entry) DecWriter() { atomic.AddInt32(&e.numWriters, -1) }
+
+// ResetSetData resets the entry and sets initial data
+func (e *Entry) ResetSetData(lists *MetricLists) {
+	atomic.StoreInt32(&e.closed, 0)
+	e.lists = lists
+	e.version = policy.InitPolicyVersion
+	e.numWriters = 0
+	e.recordLastAccessed(e.opts.ClockOptions().NowFn()())
+}
+
+// AddMetricWithPolicies adds a metric along with applicable policies
+func (e *Entry) AddMetricWithPolicies(
+	mu unaggregated.MetricUnion,
+	policies policy.VersionedPolicies,
+) error {
+	// NB(xichen): it is important that we determine the current time
+	// within the time lock. This ensures time ordering by wrapping
+	// actions that need to happen before a given time within a read lock,
+	// so it is guaranteed that actions before when a write lock is acquired
+	// must have all completed. This is used to ensure we never write metrics
+	// for times that have already been flushed.
+	timeLock := e.opts.TimeLock()
+	timeLock.RLock()
+	defer timeLock.RUnlock()
+
+	e.RLock()
+	if atomic.LoadInt32(&e.closed) == 1 {
+		e.RUnlock()
+		return errEntryClosed
+	}
+
+	currTime := e.opts.ClockOptions().NowFn()()
+	e.recordLastAccessed(currTime)
+
+	if !e.shouldUpdatePoliciesWithLock(currTime, policies.Version, policies.Cutover) {
+		e.addMetricWithLock(currTime, mu)
+		e.RUnlock()
+		return nil
+	}
+	e.RUnlock()
+
+	e.Lock()
+	if atomic.LoadInt32(&e.closed) == 1 {
+		e.Unlock()
+		return errEntryClosed
+	}
+
+	if e.shouldUpdatePoliciesWithLock(currTime, policies.Version, policies.Cutover) {
+		if err := e.updatePoliciesWithLock(mu.Type, mu.ID, policies.Policies); err != nil {
+			// NB(xichen): if an error occurred during policy update, the policies
+			// will remain as they are, i.e., there are no half-updated policies
+			e.Unlock()
+			return err
+		}
+	}
+
+	e.addMetricWithLock(currTime, mu)
+	e.Unlock()
+
+	return nil
+}
+
+// MaybeExpire attempts to expire the entry, returning true
+// if the entry has been expired and false otherwise
+func (e *Entry) MaybeExpire(now time.Time) bool {
```

Comment:
I'm not sure, we are not really trying to expire the ones that shouldn't expire. Instead, we are just testing each one to see if we we should expire and expire if needed.